File: leetcode/go/_036/main.go
package main

import "fmt"

func main() {
	fmt.Println(isValidSudoku([][]byte{
		{'5', '3', '.', '.', '7', '.', '.', '.', '.'},
		{'6', '.', '.', '1', '9', '5', '.', '.', '.'},
		{'.', '9', '8', '.', '.', '.', '.', '6', '.'},
		{'8', '.', '.', '.', '6', '.', '.', '.', '3'},
		{'4', '.', '.', '8', '.', '3', '.', '.', '1'},
		{'7', '.', '.', '.', '2', '.', '.', '.', '6'},
		{'.', '6', '.', '.', '.', '.', '2', '8', '.'},
		{'.', '.', '.', '4', '1', '9', '.', '.', '5'},
		{'.', '.', '.', '.', '8', '.', '.', '7', '9'},
	}))
}

func isValidSudoku(board [][]byte) bool {
	rows := make([]map[byte]bool, 9)
	cols := make([]map[byte]bool, 9)
	cells := make([][]map[byte]bool, 3)

	for i := 0; i < 9; i++ {
		rows[i] = make(map[byte]bool)
		cols[i] = make(map[byte]bool)
	}
	for i := 0; i < 3; i++ {
		cells[i] = make([]map[byte]bool, 3)
		for j := 0; j < 3; j++ {
			cells[i][j] = make(map[byte]bool)
		}
	}

	for i := 0; i < 9; i++ {
		for j := 0; j < 9; j++ {
			if board[i][j] == '.' {
				continue
			}
			if rows[i][board[i][j]] || cols[j][board[i][j]] || cells[i/3][j/3][board[i][j]] {
				return false
			}

			rows[i][board[i][j]] = true
			cols[j][board[i][j]] = true
			cells[i/3][j/3][board[i][j]] = true
		}
	}

	return true
}

File: leetcode/python/_036/main.py
from typing import List
from collections import defaultdict

class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        ROWS, COLS = len(board), len(board)
        rset, cset = [set() for _ in range(ROWS)], [set() for _ in range(COLS)]
        bset = [[set() for _ in range(3)] for _ in range(3)]
        for r in range(ROWS):
            for c in range(COLS):
                if board[r][c] == ".": continue
                if board[r][c] in rset[r]: return False
                if board[r][c] in cset[c]: return False
                if board[r][c] in bset[r//3][c//3]: return False
                rset[r].add(board[r][c])
                cset[c].add(board[r][c])
                bset[r//3][c//3].add(board[r][c])
        return True

print(Solution().isValidSudoku([
    ["5","3",".",".","7",".",".",".","."],
    ["6",".",".","1","9","5",".",".","."],
    [".","9","8",".",".",".",".","6","."],
    ["8",".",".",".","6",".",".",".","3"],
    ["4",".",".","8",".","3",".",".","1"],
    ["7",".",".",".","2",".",".",".","6"],
    [".","6",".",".",".",".","2","8","."],
    [".",".",".","4","1","9",".",".","5"],
    [".",".",".",".","8",".",".","7","9"]
]))